### PR TITLE
Fix IR detach after time interval

### LINKED ISSUE #4946: IR Strobe detaches after time interval
**Arma 3 Version:** `1.66` (stable)
**CBA Version:** `3.1.2` (stable)
**ACE3 Version:** `3.9.0` (stable)

**Mods:**
- `@CBA_A3`
- `@ace`

**Description:**
The time seems to be around 10 minutes.
Bux said it might be the lifeTime of the strobe. But it still continues blinking after it detaches.
I sure sounds to me like the batteries of the strobe ran out and it should stop blinking. But it continues blinking for atleast 5 minutes.

After it detaches 
`player getVariable ["ace_attach_attached",true]`
returns `[[<NULL-object>,"B_IR_Grenade"]]`

**Steps to reproduce:**
- Attach vanilla Nato IR strobe
- Go afk for 10-20 minutes
- Move a few steps away and the IR strobe will be floating in the air behind you

**Where did the issue occur?**
- Dedicated

### PR DESCRIPTION
**Pull request by Phyma**
**When merged this pull request will:**
- Will remove the x_IR_Grenade from person or vehicle when it is about to die.
- Works both for players and vehicles.
- Fixes #4946

However: Do we want it to get removed when it is about to die or just refreshed?
I think it would be one of the charms that the grenade dies after 5 minutes. While the strobe lives longer?

### CODE REVIEW

**Comment by Phyma** [comment]
Discussion about changes in: #4946 
About if it should remove or refresh.

**Comment by Phyma** [comment]
WIll post here when this is ready for review. Still need to make some changes!

**Comment by Phyma** [comment]
Ready to be reviewed!

File: addons/attach/functions/fnc_attach.sqf [hunk 1]
```diff
@@ -40,12 +40,17 @@ _onAtachText = format [localize LSTRING(Item_Attached), _onAtachText];
...
+    private _position = [0.05, -0.09, 0.1];
+    _attachedItem attachTo [_unit, _position, "leftshoulder"];
     if (!_silentScripted) then {
         _unit removeItem _itemClassname;  // Remove item
         [_onAtachText] call EFUNC(common,displayTextStructured);
     };
+
     _unit setVariable [QGVAR(attached), [[_attachedItem, _itemClassname]], true];
+
+    // Add a handler that will refresh a item when it is about to die
```
Comment (jonpas) [inline addons/attach/functions/fnc_attach.sqf]: `an item`

File: addons/attach/functions/fnc_placeApprove.sqf [hunk 2]
```diff
@@ -92,4 +92,7 @@ _attachList = _attachToVehicle getVariable [QGVAR(attached), []];
 _attachList pushBack [_attachedObject, _itemClassname];
 _attachToVehicle setVariable [QGVAR(attached), _attachList, true];
 
+// Add a handler that will refresh a item when it is about to die
```
Comment (jonpas) [inline addons/attach/functions/fnc_placeApprove.sqf]: `an item`

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 3]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Author: Phyma
+ * Remove a item of the unit when it is about to die
```
Comment (jonpas) [inline addons/attach/functions/fnc_refreshItem.sqf]: `an item from the unit` and missing final dot.

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * Arguments:
+ * 0: vehicle that it will be attached to (player or vehicle) <OBJECT>
```
Comment (jonpas) [inline addons/attach/functions/fnc_refreshItem.sqf]: Capitalize first letter, `Vehicle`, `Unit`, `Classname`.

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * 1: unit doing the attach (player) <OBJECT>
+ * 2: classname of the item you want to remove <STRING>
+ * 3: Position to be placed <ARRAY>
+ *
+ * Return Value:
+ * Nothing
```
Comment (jonpas) [inline addons/attach/functions/fnc_refreshItem.sqf]: `None`

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+#define LIFETIMEOFFSET 5
+
+params ["_attachToVehicle", "_unit", "_itemClassname", "_position"];
+
+private _itemVehClass = getText (configFile >> "CfgMagazines" >> _itemClassname >> "ACE_Attachable");
+private _timeLive = getNumber (configfile >> "CfgAmmo" >> _itemVehClass >> "timeToLive");
+
+if (0 < _timeLive && _timeLive < 1500) then {
+    //Make sure we remove  it before it is consumed
```
Comment (jonpas) [inline addons/attach/functions/fnc_refreshItem.sqf]: Double space between `remove  it`.

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+                // Add it to the list
+                _newList pushBack [_attachedObject, _itemClassname];
+            };
+        } forEach _attachedList;
+
+        // Refresh the finished list
+        _attachToVehicle setVariable [QGVAR(attached), _newList, true];
+
+    }, _timeLive, [_attachToVehicle, _unit, _itemClassname, _position]] call CBA_fnc_addPerFrameHandler;
```
Comment (jonpas) [inline addons/attach/functions/fnc_refreshItem.sqf]: This PFH never gets removed, but is started on each attach?

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+private _timeLive = getNumber (configfile >> "CfgAmmo" >> _itemVehClass >> "timeToLive");
+
+if (0 < _timeLive && _timeLive < 1500) then {
+    //Make sure we remove  it before it is consumed
+    _timeLive = _timeLive - LIFETIMEOFFSET;
+
+    [{
+        params ["_args", "_handle"];
+        _args params ["_attachToVehicle", "_unit", "_itemClassname", "_position"];
+        private ["_attachedObject"];
```
Comment (dedmen) [inline addons/attach/functions/fnc_refreshItem.sqf]: This variable is not needed outside of the forEach loop.
Also `private _attachedObject = objNull;` is probably faster.

File: addons/attach/functions/fnc_refreshItem.sqf (hunk 7 above)
Comment (PabstMirror) [inline addons/attach/functions/fnc_refreshItem.sqf]: I think we want CBA_fnc_waitAndExec.

With CBA_fnc_addPerFrameHandler, there is zero delay until first run.
e.g.
At time 20, add a PFEH with a delay of 10
First run will be at 20.01, next will be at 30.02

**Comment by Dystopian** [comment]
@Phyma could you fix this?

File: addons/attach/functions/fnc_refreshItem.sqf [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        private _attachedList = _attachToVehicle getVariable [QGVAR(attached), []];
+        private _newList = _attachedList;
+        {
+            _x params ["_xObject", "_xItemName"];
+
+            if (_xItemName == _itemClassname) then {
+                detach _xObject;
+
+                // Delete attached item after 0.5 seconds
```
Comment (dedmen) [inline addons/attach/functions/fnc_refreshItem.sqf]: This says 0.5 seconds. But code says 2 seconds.

**Comment by kymckay** [comment]
Thank you for your contribution, closing as this PR is superseded by #8283

**Pull request closed without merging**